sdk/scheduler/src/main/java/com/mesosphere/sdk/api/PodResource.java

```diff
@@ -165,6 +172,107 @@ public Response getPodInfo(@PathParam("name") String name) {
     }
 
     /**
+     * Restarts a pod in a "stopped" debug mode.
+     */
+    @Path("/{name}/stop")
+    @POST
+    public Response stopPod(@PathParam("name") String name, String bodyPayload) {
+        Set<String> taskFilter;
+        try {
+            taskFilter = parseJsonList(bodyPayload);
+        } catch (JSONException e) {
+            LOGGER.error(String.format("Failed to parse task filter '%s'", bodyPayload), e);
+            return Response.status(Response.Status.BAD_REQUEST).build();
+        }
+        try {
+            return overrideGoalState(name, taskFilter, GoalStateOverride.STOPPED);
+        } catch (Exception e) {
+            LOGGER.error(String.format("Failed to stop pod '%s' with task filter '%s'", name, taskFilter), e);
+            return Response.serverError().build();
+        }
+    }
+
+    /**
+     * Restarts a pod in a normal state following a prior "stop" command.
+     */
+    @Path("/{name}/start")
+    @POST
+    public Response startPod(@PathParam("name") String name, String bodyPayload) {
+        Set<String> taskFilter;
+        try {
+            taskFilter = parseJsonList(bodyPayload);
+        } catch (JSONException e) {
+            LOGGER.error(String.format("Failed to parse task filter '%s'", bodyPayload), e);
+            return Response.status(Response.Status.BAD_REQUEST).build();
+        }
+        try {
+            return overrideGoalState(name, taskFilter, GoalStateOverride.NONE);
+        } catch (Exception e) {
+            LOGGER.error(String.format("Failed to start pod '%s' with task filter '%s'", name, taskFilter), e);
+            return Response.serverError().build();
+        }
+    }
+
+    private static Set<String> parseJsonList(String payload) throws JSONException {
+        if (StringUtils.isBlank(payload)) {
+            return Collections.emptySet();
+        }
+        Set<String> strings = new HashSet<>();
+        Iterator<Object> iter = new JSONArray(payload).iterator();
+        while (iter.hasNext()) {
+            strings.add(iter.next().toString());
+        }
+        return strings;
+    }
+
+    private Response overrideGoalState(String podName, Set<String> taskNameFilter, GoalStateOverride override) {
+        // look up all tasks in the provided pod name:
+        List<TaskInfoAndStatus> podTasks = GroupedTasks.create(stateStore).byPod.get(podName);
```

Comment:
Can we factor out the fetching/filtering here into a separate method?

Reply:
The filtering by task name has been moved into `RequestUtils`, and the `GroupedTasks` class has been broken out into the `api.types` package.